concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,184 @@
+//! This module contains a "bump allocator", which is a simple type of allocator
+//! designed for a small code footprint and to suit most smart contracts.
+//!
+//! A bump allocator has a `next` pointer which points to the next address in
+//! the heap to hand out. The `next` pointer is incremented by the size of the
+//! each allocation. This is sufficient for the allocator to work, but it also
+//! leaks memory. However, for short-lived programs such as smart contracts,
+//! memory leaks as perfectly fine and even preferable since the need for
+//! cleanup and advanced allocation schemes is removed. The bump allocator here
+//! has two extra features:
+//!  - It keeps track of the number of active allocations in the `allocations`
+//!    field and resets the `next` pointer to `MIN_PTR_ADDR` if there are no
+//!    active allocations, thus allowing the reuse of memory.
+//!  - On deallocations it checks whether it is the very last block handed out
+//!    that is deallocated. If it is, then it moves the `next` pointer back to
+//!    the beginning of that block. Thus also allowing memory to be reused.
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    cell::UnsafeCell,
+    ptr,
+};
+
+/// The byte size of Wasm pages.
+const PAGE_SIZE: usize = 65536;
+/// FIXME: Handing out memory address 0 doesn't work, but starting with 1 does.
+const MIN_PTR_ADDR: usize = 1;
+
+/// A number of WebAssembly memory pages.
+#[derive(Eq, PartialEq, Copy, Clone)]
+struct PageCount(usize);
+
+impl PageCount {
+    /// The size in bytes. I.e. `PageCount * PAGE_SIZE`.
+    fn size_in_bytes(&self) -> usize { self.0 * PAGE_SIZE }
+}
+
+/// The number of pages returned from `memory_grow` to indicate out of memory.
+const ERROR_PAGE_COUNT: PageCount = PageCount(usize::MAX);
+
+/// This is an invalid implementation of Sync.
+/// The [BumpAllocator] cannot be safely used from multiple threads, but that is
+/// OK since it won't be in Wasm. This [Sync] implementation is required for
+/// defining it as the global allocator.
+unsafe impl Sync for BumpAllocator {}
+
+pub struct BumpAllocator {
+    /// The next pointer to be handed out.
+    next:        UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to 1 if
```

Comment:
Suggestion: In case that value changes

```suggestion
    /// The number of active allocations. Used for resetting `next` to `MIN_PTR_ADDR` if
```